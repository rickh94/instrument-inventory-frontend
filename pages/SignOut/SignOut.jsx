import React, { Component } from 'react'
import PropTypes from 'prop-types'
import {
  withStyles,
  Paper,
  FormControl,
  Input,
  InputAdornment,
  FormHelperText,
  IconButton,
  InputLabel,
  Button,
  NativeSelect,
  FormGroup,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText
} from '@material-ui/core'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBarcode } from '@fortawesome/free-solid-svg-icons'

import LoadingHeader from '../../components/LoadingHeader'
import Scanner from '../../components/Scanner'
import { root, centerStuff, lastButton, fullWidth } from '../../globalStyles'
import { API } from 'aws-amplify'

const styles = {
  root,
  centerButtons: centerStuff,
  lastButton,
  buttons: fullWidth
}

class SignOut extends Component {
  constructor(props) {
    super(props)

    this.state = {
      instrumentNumber: '',
      school: '',
      studentName: '',
      scanning: false,
      isLoading: false,
      response: '',
      errors: {}
    }
  }

  handleSubmit = async e => {
    e.preventDefault()

    if (!this.validateForm) {
      return
    }

    this.setState({ isLoading: true })

    const { instrumentNumber, school, studentName } = this.state

    try {
      const response = await API.post('instrument-inventory', 'sign-out', {
        body: {
          instrumentNumber,
          school,
          studentName
        }
      })
      this.setState({ response })
    } catch (err) {
      if (err.response.data.errors.instrumentNumber) {
        this.setState({
          errors: { instrumentNumber: err.response.data.errors.instrumentNumber }
        })
      }
      console.error(err.response)
    }
    this.setState({ isLoading: false })
  }

  validateForm = () => {
    return this.state.instrumentNumber && this.state.school && this.state.studentName
  }

  onDetected = result => {
    if (result.codeResult.code !== this.state.instrumentNumber) {
      this.setState({ instrumentNumber: result.codeResult.code, scanning: false })
    }
  }

  clearForm = () => {
    this.setState({
      instrumentNumber: '',
      school: '',
      studentName: '',
      scanning: false,
      response: ''
    })
  }

  render() {
    const { classes } = this.props
    return (
      <React.Fragment>
        <Paper className={classes.root}>
          <LoadingHeader
            isLoading={this.state.isLoading}
            title="Sign Out an Instrument"
          />
          <form onSubmit={this.handleSubmit}>
            <FormControl
              fullWidth
              error={this.state.errors.instrumentNumber ? true : false}
            >
              <InputLabel htmlFor="instrument-number">Instrument Number</InputLabel>
              {this.state.scanning ? (
                <React.Fragment>
                  <Scanner onDetected={this.onDetected} />
                  <div className={classes.centerButtons}>
                    <Button onClick={() => this.setState({ scanning: false })}>
                      Stop Scanning
                    </Button>
                  </div>
                </React.Fragment>
              ) : (
                <React.Fragment>
                  <Input
                    id="instrument-number"
                    onChange={e => this.setState({ instrumentNumber: e.target.value })}
                    aria-describedby="instrument-number-error"
                    type="text"
                    value={this.state.instrumentNumber}
                    required
                    endAdornment={
                      <InputAdornment position="end">
                        <IconButton
                          onClick={() =>
                            this.setState({ scanning: !this.state.scanning })
                          }
                        >
                          <FontAwesomeIcon icon={faBarcode} />
                        </IconButton>
                      </InputAdornment>
                    }
                  />
                  {this.state.errors && (
                    <FormHelperText id="instrument-number-error">
                      {this.state.errors.instrumentNumber}
                    </FormHelperText>
                  )}
                </React.Fragment>
              )}
            </FormControl>
            <FormControl fullWidth error={this.state.errors.studentName ? true : false}>
              <InputLabel htmlFor="student-name">Student Name</InputLabel>
              <Input
                id="student-name"
                onChange={e => this.setState({ studentName: e.target.value })}
                aria-describedby="student-name-error"
                type="text"
                value={this.state.studentName}
                required
              />
              {this.state.errors.studentName && (
                <FormHelperText id="student-name-error">
                  {this.state.errors.studentName}
                </FormHelperText>
              )}
            </FormControl>
            <FormControl fullWidth error={this.state.errors.school ? true : false}>
              <InputLabel htmlFor="school">School</InputLabel>
              <NativeSelect
                id="instrument-number"
                onChange={e => this.setState({ school: e.target.value })}
                aria-describedby="student-name-error"
                value={this.state.school}
                required
              >
                <option value="" />
                <option value="Grant Elementary School">Grant Elementary School</option>
                <option value="Hedgepath Middle School">Hedgepath Middle School</option>
                <option value="Trenton High School">Trenton High School</option>
                <option value="Columbus Elementary School">
                  Columbus Elementary School
                </option>
              </NativeSelect>
              {this.state.errors.school && (
                <FormHelperText id="student-name-error">
                  {this.state.errors.school}
                </FormHelperText>
              )}
            </FormControl>
            <FormGroup row>
              <Button onClick={this.clearForm} className={classes.lastButton}>
                Clear
              </Button>
              <Button type="submit" color="primary" disabled={!this.validateForm()}>
                Submit
              </Button>
            </FormGroup>
          </form>
        </Paper>
        <Dialog open={this.state.response ? true : false} onClose={this.clearForm}>
          <DialogContent>
            <DialogContentText>{this.state.response}</DialogContentText>
          </DialogContent>
          <DialogActions>
            <Button onClick={() => this.props.history.push('/')}>Return Home</Button>
            <Button onClick={this.clearForm}>Continue</Button>
          </DialogActions>
        </Dialog>
      </React.Fragment>
    )
  }
}

SignOut.propTypes = {
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(SignOut)
