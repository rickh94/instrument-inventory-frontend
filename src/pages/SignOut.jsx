import React, { useState, useContext, makeStyles } from 'react'
import PropTypes from 'prop-types'

import { API } from 'aws-amplify'
import { LoadingHeader, RootPaper, SchemaForm } from '../components'
import { HelpersContext } from '../contexts'
import { Button, Typography } from '@material-ui/core'
import Divider from '@material-ui/core/Divider'
import Papa from 'papaparse'
import FormControl from '@material-ui/core/FormControl'


const SignOut = ({ match, history, schema }) => {
  const { showAlert } = useContext(HelpersContext)
  const [isLoading, setLoading] = useState(false)
  const [error, setError] = useState(null)
  const [multipleData, setMultipleData] = useState()
  const [multipleErrors, setMultipleErrors] = useState('No CSV File')
  const { number } = match.params

  const handleSubmit = async body => {
    try {
      setLoading(true)
      const response = await API.post('instrument-inventory', 'sign-out', { body })
      setLoading(false)
      showAlert(response.message)
      history.push(`/instrument/${response.id}`)
    } catch (err) {
      setError(err)
      setLoading(false)
    }
  }

  const handleSubmitMultiple = async () => {
    try {
      const multipleBody = {
        instruments: multipleData,
      }
      console.log(multipleBody)
      setLoading(true)
      const response = await API.post('instrument-inventory', 'sign-out/multiple', { body: multipleBody })
      setLoading(false)
      showAlert(`Signed out ${response.updated}`)
    } catch (err) {
      setError(err)
      setLoading(false)
    }
  }

  const handleMultipleCsv = async (event) => {
    Papa.parse(event.target.files[0], {
      complete: (results, _file) => {
        if (results.data[0].number === undefined || results.data[0].assignedTo === undefined || results.data[0].location === undefined) {
          setMultipleErrors('Invalid Data')
          return
        }
        setMultipleData(results.data.filter(item => item.number.length > 3 && item.assignedTo.length > 0 && item.location.length > 0))
        setMultipleErrors(null)
      },
      header: true,
    })
  }

  return (
    <RootPaper>
      <LoadingHeader isLoading={isLoading} title="Sign Out an Instrument" />
      {schema && (
        <SchemaForm
          schema={schema.components.schemas.SignOut}
          onSubmit={handleSubmit}
          error={error}
          initialData={{ number }}
        />
      )}
      <Divider variant="middle" />
      <form onSubmit={handleSubmitMultiple} style={{ margin: '1rem auto' }}>
        <Typography variant="h6">Sign Out Multiple</Typography>
        <input type="file" id="upload-csv" accept="text/csv" style={{ display: 'none' }} onChange={handleMultipleCsv} />
        <FormControl fullWidth>
          <label htmlFor="upload-csv">
            <Button variant="contained" component="span" color="primary">Upload csv</Button>
          </label>
        </FormControl>
        {multipleErrors &&
        <Typography variant="body1">{multipleErrors}</Typography>
        }
        <Button disabled={multipleErrors}
                variant="contained"
                color="primary"
                style={{ margin: '1rem auto' }}
                onClick={handleSubmitMultiple}>Submit Multiple</Button>
      </form>
    </RootPaper>
  )
}

SignOut.propTypes = {
  schema: PropTypes.object,
  match: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
}

export default SignOut
