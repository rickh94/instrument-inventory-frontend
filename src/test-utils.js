/* eslint-disable react/prop-types */
import React from 'react'
import { MuiThemeProvider, createMuiTheme, indigo } from '@material-ui/core'
import { HelpersContext, SchemaContext } from './contexts'
import { MemoryRouter } from 'react-router-dom'
import { render } from '@testing-library/react'

const theme = createMuiTheme({
  palette: {
    primary: indigo,
    secondary: {
      main: '#0A81D1',
    },
  },
})

export const showAlert = jest.fn()
export const setSearchResults = jest.fn()
export const setFilterResults = jest.fn()
export const searchResults = jest.fn()
export const filterResults = jest.fn()

export const TestTheme = ({ children }) => (
  <MuiThemeProvider theme={theme}>{children}</MuiThemeProvider>
)

export const TestHelpers = ({ children }) => (
  <HelpersContext.Provider
    value={{
      showAlert,
      setSearchResults,
      setFilterResults,
      searchResults,
      filterResults,
    }}
  >
    {children}
  </HelpersContext.Provider>
)

export const TestRouter = ({ children }) => <MemoryRouter>{children}</MemoryRouter>

export const TestSchema = ({ children, schema = { components: { schemas: {} } } }) => (
  <SchemaContext.Provider value={schema}>{children}</SchemaContext.Provider>
)

export const TestEverything = ({ children }) => (
  <TestRouter>
    <TestTheme>
      <TestHelpers>{children}</TestHelpers>
    </TestTheme>
  </TestRouter>
)

// eslint-disable-next-line no-undef
export const flushPromises = () => new Promise(setImmediate)

const customRender = (ui, options, schema = { components: { schemas: {} } }) =>
  render(<TestSchema schema={schema}>{ui}</TestSchema>, {
    wrapper: TestEverything,
    ...options,
  })

export * from '@testing-library/react'

export { customRender as render }
