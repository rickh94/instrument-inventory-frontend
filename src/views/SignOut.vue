<template>
<div>SignOut</div>
</template>

<script>
export default {
  name: 'SignOut',
}
</script>

<style scoped>

</style>
