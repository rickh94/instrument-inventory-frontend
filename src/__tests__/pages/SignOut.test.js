import React from 'react'
import SignOut from '../../pages/SignOut'
import { render, cleanup } from '../../test-utils'

const schema = {
  components: {
    schemas: {
      SignOut: {
        title: 'SignOut',
        type: 'object',
        properties: {
          number: {
            title: 'Instrument Number',
            description: 'Instrument Number to sign out',
            type: 'string',
          },
          assignedTo: {
            title: 'Assigned To',
            description: 'Name of the Person to sign out to',
            type: 'string',
          },
          location: {
            title: 'Location',
            description: 'Primary location of instrument',
            enum: [
              'Grant Elementary School',
              'Hedgepath Middle School',
              'Wilson Elementary School',
              'Trenton High School',
              'Columbus Elementary School',
              'Office',
              'Storage',
              'Trade',
              'Maintenance',
              'Transit',
              'Unknown',
            ],
            type: 'string',
          },
        },
        required: ['number', 'assignedTo', 'location'],
      },
    },
  },
}

afterEach(cleanup)

describe('<SignOut />', () => {
  it('matches snapshot', () => {
    const { container } = render(
      <SignOut schema={schema} history={{ push: jest.fn() }} match={{ params: {} }} />
    )
    expect(container).toMatchSnapshot()
  })
})
