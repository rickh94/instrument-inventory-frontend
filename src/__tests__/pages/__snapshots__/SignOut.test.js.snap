// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignOut /> matches snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 WithStyles(ForwardRef(Paper))-root-1 MuiPaper-rounded"
  >
    <div
      class="makeStyles-heading-29"
      data-testid="loading-header"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-colorInherit"
      >
        Sign Out an Instrument
      </h5>
    </div>
    <form
      data-testid="schema-form"
      novalidate=""
    >
      <div>
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
            data-shrink="false"
            for="scanner-field"
          >
            Instrument Number
            <span
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
            data-testid="scanner-input"
            required=""
          >
            <input
              aria-describedby="scanner-error"
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
              id="scanner-field"
              required=""
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="scan-button"
                tabindex="0"
                title="Scan Barcode"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-barcode fa-w-16 "
                    data-icon="barcode"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 512 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M0 448V64h18v384H0zm26.857-.273V64H36v383.727h-9.143zm27.143 0V64h8.857v383.727H54zm44.857 0V64h8.857v383.727h-8.857zm36 0V64h17.714v383.727h-17.714zm44.857 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm18 0V64h8.857v383.727h-8.857zm35.715 0V64h18v383.727h-18zm44.857 0V64h18v383.727h-18zm35.999 0V64h18.001v383.727h-18.001zm36.001 0V64h18.001v383.727h-18.001zm26.857 0V64h18v383.727h-18zm45.143 0V64h26.857v383.727h-26.857zm35.714 0V64h9.143v383.727H476zm18 .273V64h18v384h-18z"
                      fill="currentColor"
                    />
                  </svg>
                  <span
                    class="MuiTypography-root MuiTypography-srOnly"
                    data-testid="sr-text"
                  >
                    Scan Barcode
                  </span>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
          description="Name of the Person to sign out to"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-required Mui-required"
            data-shrink="false"
            for="uniforms-0000-0003"
          >
            Assigned To
            <span
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              id="uniforms-0000-0003"
              name="assignedTo"
              placeholder=""
              required=""
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense Mui-required Mui-required"
            data-shrink="false"
          >
            Location
            <span
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            appearance="checkbox"
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense"
            description="Primary location of instrument"
          >
            <div
              aria-haspopup="listbox"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              id="select-location"
              role="button"
              tabindex="0"
            >
              <span>
                ​
              </span>
            </div>
            <input
              id="uniforms-0000-0005"
              name="location"
              type="hidden"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="MuiFormGroup-root MuiFormGroup-row"
      >
        <button
          class="MuiButtonBase-root MuiButton-root makeStyles-button-61 makeStyles-firstButton-63 makeStyles-lastButton-62 MuiButton-text MuiButton-textPrimary"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Submit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </div>
</div>
`;
